File: 06-DevOps/02-CI-CD/制品晋升与语义化版本/go/semver2.go
// SemVer 2.0.0 解析与优先级比较(Go 对照实现)。零第三方依赖, 仅标准库。
//
// 权威依据: semver.org 规范原文(semver/semver 仓库的 semver.md, 本 demo 实读):
// 前导零非法、预发布低于正规版本、构建元数据不参与比较、数字标识符恒低于非数字、
// 前缀相等时标识符多者优先级更高。官方给的带编号捕获组正则原样使用。
package main

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

// semverRE 是官方给出的「带编号捕获组」正则(ECMA/PCRE/Python/Go 通用)。
var semverRE = regexp.MustCompile(
	`^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)` +
		`(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)` +
		`(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?` +
		`(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$`)

// Version 是一个合法的语义化版本;Build 不参与优先级比较。
type Version struct {
	Major int
	Minor int
	Patch int
	Pre   []string
	Build []string
}

// ParseVersion 解析并校验;非法版本一律报错(不静默降级)。
func ParseVersion(text string) (Version, error) {
	m := semverRE.FindStringSubmatch(text)
	if m == nil {
		return Version{}, fmt.Errorf("不是合法的 SemVer 2.0.0: %q", text)
	}
	major, err1 := strconv.Atoi(m[1])
	minor, err2 := strconv.Atoi(m[2])
	patch, err3 := strconv.Atoi(m[3])
	if err1 != nil || err2 != nil || err3 != nil {
		return Version{}, fmt.Errorf("版本号数值溢出: %q", text)
	}
	v := Version{Major: major, Minor: minor, Patch: patch}
	if m[4] != "" {
		v.Pre = strings.Split(m[4], ".")
	}
	if m[5] != "" {
		v.Build = strings.Split(m[5], ".")
	}
	return v, nil
}

// MustVersion 供断言使用: 解析失败直接 panic, 让测试立刻炸出来。
func MustVersion(text string) Version {
	v, err := ParseVersion(text)
	if err != nil {
		panic(err)
	}
	return v
}

func (v Version) String() string {
	out := fmt.Sprintf("%d.%d.%d", v.Major, v.Minor, v.Patch)
	if len(v.Pre) > 0 {
		out += "-" + strings.Join(v.Pre, ".")
	}
	if len(v.Build) > 0 {
		out += "+" + strings.Join(v.Build, ".")
	}
	return out
}

// IsPrerelease 报告是否有预发布标识符。
func (v Version) IsPrerelease() bool { return len(v.Pre) > 0 }

// Compare 按官方优先级规则返回 -1/0/1;构建元数据不参与。
func (v Version) Compare(o Version) int {
	if c := cmpInt(v.Major, o.Major); c != 0 {
		return c
	}
	if c := cmpInt(v.Minor, o.Minor); c != 0 {
		return c
	}
	if c := cmpInt(v.Patch, o.Patch); c != 0 {
		return c
	}
	if len(v.Pre) == 0 && len(o.Pre) == 0 {
		return 0
	}
	// 有预发布的一方优先级更低
	if len(v.Pre) == 0 {
		return 1
	}
	if len(o.Pre) == 0 {
		return -1
	}
	n := len(v.Pre)
	if len(o.Pre) < n {
		n = len(o.Pre)
	}
	for i := 0; i < n; i++ {
		if c := CompareIdentifier(v.Pre[i], o.Pre[i]); c != 0 {
			return c
		}
	}
	// 前缀全相等: 标识符更多的一方优先级更高
	return cmpInt(len(v.Pre), len(o.Pre))
}

// CompareIdentifier 实现官方标识符比较: 纯数字按数值; 数字恒低于非数字; 否则 ASCII 序。
func CompareIdentifier(a, b string) int {
	aNum, bNum := isDigits(a), isDigits(b)
	switch {
	case aNum && bNum:
		ai, _ := strconv.Atoi(a)
		bi, _ := strconv.Atoi(b)
		return cmpInt(ai, bi)
	case aNum:
		return -1
	case bNum:
		return 1
	}
	return strings.Compare(a, b)
}

// Bump 按官方递增规则产生下一个版本;递增时清空预发布与构建元数据。
func (v Version) Bump(kind string) (Version, error) {
	switch kind {
	case "major":
		return Version{Major: v.Major + 1}, nil
	case "minor":
		return Version{Major: v.Major, Minor: v.Minor + 1}, nil
	case "patch":
		return Version{Major: v.Major, Minor: v.Minor, Patch: v.Patch + 1}, nil
	case "prerelease":
		// 本规范未规定预发布递增;这里是工程约定: 下一个 patch 的 -0
		return Version{Major: v.Major, Minor: v.Minor, Patch: v.Patch + 1,
			Pre: []string{"0"}}, nil
	}
	return Version{}, fmt.Errorf("未知的递增类型: %s", kind)
}

// Highest 返回优先级最高的版本;includePrerelease 为 false 时忽略预发布。
func Highest(texts []string, includePrerelease bool) (Version, bool) {
	var best Version
	found := false
	for _, t := range texts {
		v, err := ParseVersion(t)
		if err != nil {
			continue
		}
		if v.IsPrerelease() && !includePrerelease {
			continue
		}
		if !found || v.Compare(best) > 0 {
			best, found = v, true
		}
	}
	return best, found
}

func cmpInt(a, b int) int {
	switch {
	case a > b:
		return 1
	case a < b:
		return -1
	}
	return 0
}

// isDigits 只认 ASCII 0-9(与规范 identifier 字符集一致)。
func isDigits(s string) bool {
	if s == "" {
		return false
	}
	for _, r := range s {
		if r < '0' || r > '9' {
			return false
		}
	}
	return true
}

File: 06-DevOps/02-CI-CD/制品晋升与语义化版本/python/semver2.py
"""SemVer 2.0.0 解析与优先级比较(可执行)。零第三方依赖。

权威依据: semver.org 规范原文(仓库 semver/semver 的 semver.md, 本 demo 实读)。

核心事实(全部有原文支撑):
1. 正规版本号形如 ``X.Y.Z``,X/Y/Z 是非负整数且**不得有前导零**;
   ``1.9.0 -> 1.10.0`` 是**数值**递增, 不是字符串排序。
2. ``-`` 后的点分标识符是**预发布**;标识符只能是 ASCII 字母数字与连字符,
   不得为空, 且**数字标识符不得有前导零**。预发布版本优先级**低于**对应正规版本。
3. ``+`` 后是**构建元数据**;比较优先级时**必须忽略**它 —— 只差构建元数据的两个版本
   优先级相同(``1.0.0+a`` 与 ``1.0.0+b`` 不分先后)。
4. 优先级逐段左到右比较:先 major/minor/patch(**数值**)比;若相等,有预发布 < 无预发布;
   再逐段比预发布标识符 —— 纯数字**按数值**比、含字母或连字符**按 ASCII 字典序**比、
   **数字标识符恒低于非数字标识符**;前缀全相等时**标识符多的优先级更高**。
   官方例子: ``1.0.0-alpha < 1.0.0-alpha.1 < 1.0.0-alpha.beta < 1.0.0-beta
   < 1.0.0-beta.2 < 1.0.0-beta.11 < 1.0.0-rc.1 < 1.0.0``。
5. ``v1.2.3`` **不是**语义化版本(FAQ 明确): ``v`` 只是 tag 名的前缀习惯。
6. 已发布的版本其内容不得修改(与 OCI 的 digest 不可变语义同构)。
"""

from __future__ import annotations

import re

# 官方给出的「带编号捕获组」正则(ECMA/PCRE/Python/Go 通用), 原样照抄 semver.md。
SEMVER_RE = re.compile(
    r"^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)"
    r"(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)"
    r"(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?"
    r"(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$"
)


class SemverError(ValueError):
    """版本号不符合 SemVer 2.0.0(解析期错误, 不能静默降级)。"""


class Version:
    """一个合法的语义化版本。``build`` 不参与优先级比较。"""

    __slots__ = ("major", "minor", "patch", "prerelease", "build")

    def __init__(self, major, minor, patch, prerelease=(), build=()):
        self.major = major
        self.minor = minor
        self.patch = patch
        self.prerelease = tuple(prerelease)
        self.build = tuple(build)

    # ------------------------------------------------------------- 解析
    @classmethod
    def parse(cls, text: str) -> "Version":
        if not isinstance(text, str):
            raise SemverError("版本号必须是字符串")
        m = SEMVER_RE.match(text)
        if not m:
            raise SemverError("不是合法的 SemVer 2.0.0: %r" % text)
        major, minor, patch, pre, build = m.groups()
        return cls(int(major), int(minor), int(patch),
                   pre.split(".") if pre else (),
                   build.split(".") if build else ())

    # ------------------------------------------------------------- 输出
    @property
    def core(self) -> str:
        return "%d.%d.%d" % (self.major, self.minor, self.patch)

    def __str__(self) -> str:
        out = self.core
        if self.prerelease:
            out += "-" + ".".join(self.prerelease)
        if self.build:
            out += "+" + ".".join(self.build)
        return out

    def __repr__(self) -> str:
        return "Version(%s)" % self

    @property
    def is_prerelease(self) -> bool:
        return bool(self.prerelease)

    # ------------------------------------------------------------- 比较
    def compare(self, other: "Version") -> int:
        """按官方 11 条优先级规则返回 -1/0/1;构建元数据不参与。"""
        for a, b in ((self.major, other.major), (self.minor, other.minor),
                     (self.patch, other.patch)):
            if a != b:
                return 1 if a > b else -1
        if self.prerelease == other.prerelease:
            return 0
        # 有预发布的一方优先级更低
        if not self.prerelease:
            return 1
        if not other.prerelease:
            return -1
        for a, b in zip(self.prerelease, other.prerelease):
            c = _compare_identifier(a, b)
            if c:
                return c
        # 前缀全相等: 标识符更多的一方优先级更高
        if len(self.prerelease) == len(other.prerelease):
            return 0
        return 1 if len(self.prerelease) > len(other.prerelease) else -1

    def __eq__(self, other) -> bool:
        return isinstance(other, Version) and self.compare(other) == 0

    def __lt__(self, other) -> bool:
        return self.compare(other) < 0

    def __le__(self, other) -> bool:
        return self.compare(other) <= 0

    def __gt__(self, other) -> bool:
        return self.compare(other) > 0

    def __ge__(self, other) -> bool:
        return self.compare(other) >= 0

    def __hash__(self) -> int:
        # 与 __eq__ 保持一致: 只差构建元数据的版本必须同哈希
        return hash((self.major, self.minor, self.patch, self.prerelease))

    # ------------------------------------------------------------- 递增
    def bump(self, kind: str) -> "Version":
        """按官方递增规则产生下一个版本(递增时**清空**预发布与构建元数据)。"""
        if kind == "major":
            return Version(self.major + 1, 0, 0)
        if kind == "minor":
            return Version(self.major, self.minor + 1, 0)
        if kind == "patch":
            return Version(self.major, self.minor, self.patch + 1)
        if kind == "prerelease":
            # 预发布递增是本规范未规定的工程约定: 1.2.3 -> 1.2.4-0
            return Version(self.major, self.minor, self.patch + 1, ("0",))
        raise SemverError("未知的递增类型: %s" % kind)


def _compare_identifier(a: str, b: str) -> int:
    """官方规则: 纯数字按数值比; 数字恒低于非数字; 否则 ASCII 字典序。"""
    a_num = a.isdigit() and a.isascii()
    b_num = b.isdigit() and b.isascii()
    if a_num and b_num:
        ai, bi = int(a), int(b)
        return 0 if ai == bi else (1 if ai > bi else -1)
    if a_num:
        return -1
    if b_num:
        return 1
    return 0 if a == b else (1 if a > b else -1)


def sort_versions(texts) -> list:
    """把版本字符串按优先级升序排列(用插入排序展示比较次数以外的语义)。"""
    out = []
    for t in texts:
        v = Version.parse(t)
        pos = len(out)
        while pos > 0 and out[pos - 1].compare(v) > 0:
            pos -= 1
        out.insert(pos, v)
    return out


def highest(texts, include_prerelease: bool = True):
    """返回优先级最高的版本;``include_prerelease=False`` 时忽略预发布版本。"""
    best = None
    for t in texts:
        v = Version.parse(t)
        if v.is_prerelease and not include_prerelease:
            continue
        if best is None or v.compare(best) > 0:
            best = v
    return best
